smartsim/_core/mli/infrastructure/control/requestdispatcher.py

```diff
@@ -0,0 +1,515 @@
+            raise Full
+        try:
+            if self.full():
+                raise Full
+            if self._first_put is None:
+                self._first_put = time.time()
+            super().put(item, block=block, timeout=timeout)
+        finally:
+            self.release()
+
+    @property
+    def _elapsed_time(self) -> float:
+        if self.empty():
+            return 0
+        return time.time() - self._first_put
+
+    @property
+    def ready(self) -> bool:
+        """True if the queue can be flushed"""
+        if self.empty():
+            return False
+        return self.full() or (self._elapsed_time >= self._batch_timeout)
+
+    def make_disposable(self) -> None:
+        """Set this queue as disposable, and never use it again after it gets flushed"""
+        self._disposable = True
+
+    @property
+    def can_be_removed(self) -> bool:
+        """Whether this queue can be deleted and garbafe collected"""
+        return self.empty() and self._disposable
+
+    def flush(self) -> list[t.Any]:
+        """Get all requests from queue
+        :return: Requests waiting to be executed
+        """
+        num_items = self.qsize()
+        self._first_put = None
+        items = []
+        for _ in range(num_items):
+            try:
+                items.append(self.get())
+            except Empty:
+                break
+
+        return items
+
+    def full(self) -> bool:
+        """Return True if the queue has reached its maximum capacity"""
+        if self._disposable:
+            return True
+        if self._batch_size <= 0:
+            return False
+        return self.qsize() >= self._batch_size
+
+    def empty(self) -> bool:
+        """Return True if the queue has 0 elements"""
+        return self.qsize() == 0
+
+
+class RequestDispatcher(Service):
+    def __init__(
+        self,
+        batch_timeout: float,
+        batch_size: int,
+        config_loader: EnvironmentConfigLoader,
+        worker_type: t.Type[MachineLearningWorkerBase],
+        comm_channel_type: t.Type[CommChannelBase] = DragonCommChannel,
+    ) -> None:
+        """The RequestDispatcher intercepts inference requests, stages them in
+        queues and batches them together before making them available to Worker
+        Managers.
+        :param batch_timeout: Maximum elapsed time before flushing a complete or incomplete batch
+        :param batch_size: Total capacity of each batch queue.
+        :param mem_pool: Memory pool used to share batched input tensors with worker
+        managers
+        :param config_loader: Object to load configuration from environment
+        :param worker_type: Type of worker to instantiate to batch inputs
+        :param comm_channel_type: Type of channel used to get requests
+        :raises SmartSimError: If config_loaded.get_queue() does not return a channel
+        """
+        super().__init__(as_service=True, cooldown=1)
+        self._queues: dict[str, list[BatchQueue]] = []
```

Comment:
Calling this out because this might be affected by the typedef I proposed before of ``ModelIdentifier``

Reply:
I will keep `str` right now, as it should only boil down to the model name, not its ID/version -- there must be just one queue per model name.